File: backend/app/api/v1/endpoints/cases.py
"""Case endpoints."""

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.api.deps import get_db
from app.schemas.case import CaseCreate, CaseListResponse, CaseRead
from app.services import case_service
from app.services.errors import ConflictError, NotFoundError

router = APIRouter()


@router.post("", response_model=CaseRead, status_code=status.HTTP_201_CREATED)
def create_case(payload: CaseCreate, db: Session = Depends(get_db)):
    """Create a case."""
    try:
        return case_service.create_case(db, payload)
    except ConflictError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc)) from exc


@router.get("", response_model=CaseListResponse)
def list_cases(
    limit: int = Query(default=100, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
    db: Session = Depends(get_db),
):
    """List cases."""
    return {"items": case_service.list_cases(db, limit=limit, offset=offset)}


@router.get("/{case_id}", response_model=CaseRead)
def get_case(case_id: UUID, db: Session = Depends(get_db)):
    """Get one case."""
    try:
        return case_service.get_case(db, case_id)
    except NotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc

File: frontend/src/api/cases.ts
import { apiRequest, jsonBody } from "./client";
import type { Case, ListResponse } from "../types/domain";

export interface CaseCreatePayload {
  case_code: string;
  name: string;
  description?: string | null;
  status?: string;
}

export interface CaseListParams {
  limit?: number;
  offset?: number;
}

export function listCases(params: CaseListParams = {}): Promise<ListResponse<Case>> {
  const query = new URLSearchParams();
  query.set("limit", String(params.limit ?? 50));
  query.set("offset", String(params.offset ?? 0));
  return apiRequest<ListResponse<Case>>(`/api/v1/cases?${query.toString()}`);
}

export function getCase(caseId: string): Promise<Case> {
  return apiRequest<Case>(`/api/v1/cases/${caseId}`);
}

export function createCase(payload: CaseCreatePayload): Promise<Case> {
  return apiRequest<Case>("/api/v1/cases", {
    method: "POST",
    body: jsonBody(payload),
  });
}
